pkg/api/staking.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2022 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package api
+
+import (
+	"errors"
+	"math/big"
+	"net/http"
+	"strconv"
+
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/sctx"
+	"github.com/ethersphere/bee/pkg/staking/stakingcontract"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/gorilla/mux"
+)
+
+func (s *Service) stakingAccessHandler(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if !s.stakingSem.TryAcquire(1) {
+			s.logger.Debug("staking access: simultaneous on-chain operations not supported")
+			s.logger.Error(nil, "staking access: simultaneous on-chain operations not supported")
+			jsonhttp.TooManyRequests(w, "simultaneous on-chain operations not supported")
+			return
+		}
+		defer s.stakingSem.Release(1)
+
+		h.ServeHTTP(w, r)
+	})
+}
+
+type getStakeResponse struct {
+	StakedAmount *big.Int `json:"stakedAmount"`
+}
+
+func (s *Service) stakingDepositHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	if price, ok := r.Header[gasPriceHeader]; ok {
+		p, ok := big.NewInt(0).SetString(price[0], 10)
+		if !ok {
+			s.logger.Error(nil, "deposit stake: bad gas price")
+			jsonhttp.BadRequest(w, errBadGasPrice)
+			return
+		}
+		ctx = sctx.SetGasPrice(ctx, p)
+	}
+
+	if limit, ok := r.Header[gasLimitHeader]; ok {
+		l, err := strconv.ParseUint(limit[0], 10, 64)
+		if err != nil {
+			s.logger.Error(err, "deposit stake: bad gas limit")
+			jsonhttp.BadRequest(w, errBadGasLimit)
+			return
+		}
+		ctx = sctx.SetGasLimit(ctx, l)
+	}
```

Comment:
could we set `GasLimit` and `GasPrice` to request's context on the middleware level? this way we wouldn't need to do this for each and every handler

Reply:
yes we can but since the scope of this task is related to staking agent i think it would be better if we use this and then have a separate PR for that if needed. That being said there is another pr which is focusing on this issue so maybe it would be better to have this suggestion implemented there since that is focused on exact this issue.

Reply:
In that case I would suggest not committing code related to setting gas price and gas limit to context in this PR. And wait for other PR to implement this.

Reply:
this is fine for now, let's merge as is, and create a PR for the middleware in another PR. 

there are other endpoints that needs the same treatment as well